docs: improve contributing and pull request documentation (#5387)
adobe/spectrum-web-components: CONTRIBUTING.md

-   [1. Community \& Support](#1-community--support)
    -   [1.1. External Contributors](#11-external-contributors)
    -   [1.2. Internal Contributors](#12-internal-contributors)
-   [2. Ways to Contribute](#2-ways-to-contribute)
-   [3. Contributor License Agreement](#3-contributor-license-agreement)
-   [4. Code of Conduct](#4-code-of-conduct)
-   [5. Using the Issue Tracker](#5-using-the-issue-tracker)
    -   [5.1. Before Submitting A Bug Report](#51-before-submitting-a-bug-report)
-   [6. Bug Reports](#6-bug-reports)
    -   [6.1. Issue Severity Classification](#61-issue-severity-classification)
-   [7. Feature Requests/New Component](#7-feature-requestsnew-component)
-   [8. Pull Requests](#8-pull-requests)
-   [9. Branches](#9-branches)
-   [10. Developing Locally](#10-developing-locally)
-   [11. Testing](#11-testing)
-   [12. Documentation](#12-documentation)
-   [13. Best Practices \& Guidelines](#13-best-practices--guidelines)
    -   [13.1. Code Formatting](#131-code-formatting)
    -   [13.2. Accessibility](#132-accessibility)
    -   [13.3. Commit Guidelines](#133-commit-guidelines)
-   [14. Thank You](#14-thank-you)

---

# 1. Community & Support

> A fantastic first step to contributing is joining our community discussions. This is where you can:

Author: this is definitely something we should discuss. i know in the past we used discussions but we can also remove this until we are ready to commit.

Reviewer: Yeah, I think actually ppl opening an issue is better??? Since we no longer use JUST GH issues for triaging

Author: want to take a stab an language around this?

Reviewer 2: Agree with najika, even if we route them to start a discussion, users feel comfortable to discuss in the channel itself and its understandable

Author: i was trying to keep the language "open source" focused but we can separate our users concerns of slack for internal users versus github issues for external users. 

Feel free to make a code suggestions to improve this section

> A fantastic first step to contributing is joining our community discussions. This is where you can:
> 
> -   Ask questions and troubleshoot with other users.
> -   Propose new ideas or get feedback on your own.
> -   Stay informed about what’s coming next.
> 
> ## 1.1. External Contributors
> 
> If you need support or have a question about how something works, GitHub Discussions is the best place to start.

Reviewer: ```suggestion
A fantastic first step to contributing is filing an issue. This is where you can:
-   Ask questions, file bugs, and troubleshoot with other users.
-   Propose new features and ideas or get feedback on your own through a linked pull request.

Additionally, you can check GitHub Discussions to stay up-to-date with any major announcements about the project. 

## 1.1. External Contributors

**Adobe Employees, read section 1.2 below.** 

If you need support or have a question about how something works, filing an issue is the best place to start. A team member will be in touch to either triage your issue or follow-up with you in the comments.
```

## 1.2. Internal Contributors

If you work for Adobe, our slack channel #spectrum_web_components has some great workflows to get you started. Be sure to read the Welcome Canvas when you join.

---

> # 2. Ways to Contribute

Reviewer 2: ```suggestion
# 2. How can you Contribute
```

We require all external contributors to sign our Contributor License Agreement (CLA). If you haven’t signed it before making your first contribution, please do so—otherwise, we can’t merge your changes.

---

# 4. Code of Conduct

Spectrum Web Components abides by the Adobe Code of Conduct. By participating, you agree to treat all community members kindly and respectfully. We’re committed to fostering a welcoming, inclusive environment.
Should any behavior fall short of these expectations, please report it to <Grp-opensourceoffice@adobe.com>.

---

# 5. Using the Issue Tracker

We use GitHub Issues for two purposes:

1. Bug Reports
2. Feature Requests (after initial discussion)

If you’re having a usage issue or need support, do not open an issue. Instead, reference the Community & Support section. This helps us keep issues focused on actual bugs and actionable tasks.

## 5.1. Before Submitting A Bug Report

1. Check the Open Issues to see if the problem has already been reported.
    1. TIP: Apply the component label to make your search process more straightforward.
2. If it has and the issue is still open, add a comment to the existing issue instead of opening a new one.

> 3. Check if you can reproduce the problem in the latest version of Spectrum Web Components.

Author: we should append this with our new isolation steps in another section once we land on a new tool

---

# 7. Feature Requests/New Component

Is there something you wish the project did differently? Have a new component in mind? We love hearing new ideas and are eager to collaborate!

-   Start with a discussion: Share your idea in Discussions to gather feedback and see if it aligns with project goals.
-   Open a feature request issue: After some positive initial conversation, open an issue using the `Feature Request` or `New Component` template with details and potential use cases.

---

# 8. Pull Requests

If you plan to fix a bug, create a feature, or improve documentation, follow the [Pull Request Guide](PULL_REQUESTS.md) to ensure you're contribution meets expectations for getting reviewed.

---

# 9. Branches

> We keep things straightforward with branching:

Author: I can get on board with that!! great call out!! love leading with empathy!

-   `[username]/[short-description]` (e.g., `alex/fix-dropdown-bug`) is often all you need.
-   If your work references a known issue, you could also incorporate the issue number (e.g., `alex/123-bug-fix`).

> Avoid editing distribution files (if present). Make changes to the source files, then allow the build system to generate any bundled or output files automatically.

Author: we can update this line after cutoff

---

# 10. Developing Locally

> Read the steps outlined in the README.md to get your environment set up.

Author: ```suggestion
Read the steps outlined in the [README.md](README.md) to get your environment set up.
```

---

# 11. Testing

Quality and stability are important. We require writing tests for any fixes or features you introduce. This helps ensure:

-   Bugs don’t resurface later.
-   New features work as intended for all users.
-   Overall library reliability remains high.

Read about our testing guidance in the README.md.

If you’re unsure how to write tests for certain parts of the library, don’t hesitate to ask maintainers for guidance. We appreciate every effort to keep the code solid!

---

# 12. Documentation

In addition to well-tested code, documentation is crucial. Whenever you add or change a feature,include documentation for it in the relevant areas:

-   **README.md**: Each component has a README within its directory. Ensure your changes are included here. This file is used in our generated documentation site.
-   **Comment Annotations**: We use comment-based documentation (JSDocs) so that references are generated automatically where possible.

Accessible, helpful docs are a huge win for everyone, especially newcomers.

---

# 13. Best Practices & Guidelines

## 13.1. Code Formatting

> We rely on automated tools like Prettier or ESLint to enforce style preferences. Setting up these tools in your editor saves time and prevents minor style conflicts from slowing down reviews.

Author: i would love to improve this section and improved out extensions.json file and link it here. would you be cool with a follow up ticket to improve this later?

## 13.2. Accessibility

Since Spectrum Web Components is used by a diverse audience, we pay close attention to accessibility. Please keep features inclusive by:

-   Using semantic markup when possible.
-   Properly labeling interactive elements.
-   Ensuring focus states and keyboard navigation are well supported.

If you’re unsure about an accessibility detail, open a discussion or ask in your PR.

## 13.3. Commit Guidelines

> As mentioned previously, we use Conventional Commit syntax:

Author: we can but for the time being we keep this until that conversation happens and we update this at that time

-   Start with a discussion: Share your idea in Discussions to gather feedback and see if it aligns with project goals.
-   Open a feature request issue: After some positive initial conversation, open an issue using the `Feature Request` or `New Component` template with details and potential use cases.

```bash
type(scope?): subject #scope is optional, but should reference the package you are updating
```
---

We maintain a specific standard of commit messages while committing to the respository. See [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
# 8. Pull Requests

## Browser support
If you plan to fix a bug, create a feature, or improve documentation, follow the [Pull Request Guide](PULL_REQUESTS.md) to ensure you're contribution meets expectations for getting reviewed.

When contributing to our system or filing an issue, please consider the target browsers we support.
---

We support the latest 2 major versions of these browsers for desktop:
# 9. Branches

> -   Google Chrome

Author: this is captured in the README: https://github.com/adobe/spectrum-web-components/blob/5950185df5d30694951dd74f07d12168de18bb12/README.md#L7

-   [Testing](testing)
-   [Documentation](documentation)
-   [Best Practices \& Guidelines](best-practices--guidelines)
    -   [Code Formatting](code-formatting)
    -   [Accessibility](accessibility)
    -   [Commit Guidelines](commit-guidelines)
-   [Thank You](thank-you)

---

## Community & Support

A fantastic first step to contributing is filing an issue. This is where you can:

-   Ask questions, file bugs, and troubleshoot with other users.
-   Propose new features and ideas or get feedback on your own through a linked pull request.
-   Additionally, you can check GitHub Discussions to stay up-to-date with any major announcements about the project.

### External Contributors

**Adobe Employees, read Internal Contributors section below.**
If you need support or have a question about how something works, filing an issue is the best place to start. A team member will be in touch to either triage your issue or follow-up with you in the comments.

### Internal Contributors

If you work for Adobe, our Slack channel #spectrum_web_components has some great workflows to get you started. Be sure to read the Welcome Canvas when you join.

---

> ## How can you Contribute

Reviewer 3: `How you can contribute` reads better to me, but just a nit.

---

## Bug reports

When you file a bug, please use the `Bug Report` template provided in GitHub. Include the following information:

-   A concise summary of the problem.
-   Relevant components involved in the issue.
-   Issue Severity based on our classifications defined below.
-   What you expected vs. what actually happened, along with any errors logged in the console.
-   Steps to reproduce the issue, preferably in an isolated environment, so that we can narrow down where the bug is originating from. (e.g., webcomponents.dev or CodePen). Be detailed if you write out the steps!
-   Relevant environment details (OS, browser, library version).

Clear bug reports speed up the triage process, help us replicate the issue, and keep the project robust.

### Issue severity classification

Providing the correct issue severity classification helps us adequately assess and prioritize your issue. We reserve the right to adjust the severity of your bug during triage.
Below is our issue severity classification criteria:

> | Severity Level | Description                                                                                                                                                                                                                                                                                                                                                     | Examples                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |

Author: ```suggestion
| Severity Level      | Description                                                                                                                                                                                                                                                                                                                                                     | Examples                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
```

> | -------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

Author: ```suggestion
| ------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
```

> | 🔥 SEV 1       | A critical design or functionality issue that breaks the design system, causes significant usability problems, or exposes critical security vulnerabilities. This issue impacts all users and/or essential workflows, making users unable to complete any tasks. Requires immediate attention.                                                                  | Broken navigation system, complete unresponsiveness on all devices, components not rendering, inaccessible primary actions, security vulnerabilities.<br><br>Accessibility: The end user is not able to complete core tasks or activities (e.g., key navigational elements not accessible via keyboard, missing or incorrect form labels that prevent screen reader users from completing forms or actions, critical color contrast issues that prevent users from reading or interacting with essential content). |

Author: ```suggestion
| 🔥 SEV 1.           | A critical design or functionality issue that breaks the design system, causes significant usability problems, or exposes critical security vulnerabilities. This issue impacts all users and/or essential workflows, making users unable to complete any tasks. Requires immediate attention.                                                                  | Broken navigation system, complete unresponsiveness on all devices, components not rendering, inaccessible primary actions, security vulnerabilities.<br><br>Accessibility: The end user is not able to complete core tasks or activities (e.g., key navigational elements not accessible via keyboard, missing or incorrect form labels that prevent screen reader users from completing forms or actions, critical color contrast issues that prevent users from reading or interacting with essential content). |
```

> | 🔴 SEV 2       | A significant problem affecting the design, usability, or functionality of the system or components for a subset of users. This issue causes major disruptions and prevent users from completing tasks or workflows.<br><br>Accessibility: Does not conform with WCAG 2.1 Level AA criteria and blocks core user tasks (no known workaround).                   | Content that is out of view or unreachable by customers, critical buttons or links not functioning, or actions that cannot be submitted. Unintentional breaking changes causing disruptions in production environments.<br><br>Accessibility: Issue with a component or controller with widespread use and blocks core user tasks (no known workaround).                                                                                                                                                           |

Author: ```suggestion
| 🔴 SEV 2            | A significant problem affecting the design, usability, or functionality of the system or components for a subset of users. This issue causes major disruptions and prevent users from completing tasks or workflows.<br><br>Accessibility: Does not conform with WCAG 2.1 Level AA criteria and blocks core user tasks (no known workaround).                   | Content that is out of view or unreachable by customers, critical buttons or links not functioning, or actions that cannot be submitted. Unintentional breaking changes causing disruptions in production environments.<br><br>Accessibility: Issue with a component or controller with widespread use and blocks core user tasks (no known workaround).                                                                                                                                                           |
```

> | 🟠 SEV 3       | A design or functionality issue that causes noticeable errors or minor usability problems for users which either cause confusion or degrade the user experience, but do not prevent task completion.<br><br>Accessibility: Does not conform with WCAG 2.1 Level AA criteria and is either non-blocking for core user tasks or blocking for non-core user tasks. | Misleading labels, inconsistent component behavior, unexpected interactions, decreases in system performance.<br><br>Accessibility: Workarounds are available and discoverable for the end user to complete core user tasks, or the end user is not able to complete non-core user tasks (e.g., inadequate ARIA labels, improper focus management, insufficient color contrast for non-critical elements).                                                                                                         |

Author: ```suggestion
| 🟠 SEV 3            | A design or functionality issue that causes noticeable errors or minor usability problems for users which either cause confusion or degrade the user experience, but do not prevent task completion.<br><br>Accessibility: Does not conform with WCAG 2.1 Level AA criteria and is either non-blocking for core user tasks or blocking for non-core user tasks. | Misleading labels, inconsistent component behavior, unexpected interactions, decreases in system performance.<br><br>Accessibility: Workarounds are available and discoverable for the end user to complete core user tasks, or the end user is not able to complete non-core user tasks (e.g., inadequate ARIA labels, improper focus management, insufficient color contrast for non-critical elements).                                                                                                         |
```

> | 🟡 SEV 4       | A minor design flaw that affects user experience, but doesn't have a serious impact on overall usability or functionality. This issue does not prevent users from completing tasks.<br><br>Accessibility: Does not conform with WCAG 2.1 Level AA criteria but has lower user impact.                                                                           | Minor visual inconsistencies, non-critical content misalignment, or minor layout issues.<br><br>Accessibility: A WCAG violation is present in areas not frequently visited by end users, or it has a lower impact on overall accessibility and usability.                                                                                                                                                                                                                                                          |

Author: ```suggestion
| 🟡 SEV 4            | A minor design flaw that affects user experience, but doesn't have a serious impact on overall usability or functionality. This issue does not prevent users from completing tasks.<br><br>Accessibility: Does not conform with WCAG 2.1 Level AA criteria but has lower user impact.                                                                           | Minor visual inconsistencies, non-critical content misalignment, or minor layout issues.<br><br>Accessibility: A WCAG violation is present in areas not frequently visited by end users, or it has a lower impact on overall accessibility and usability.                                                                                                                                                                                                                                                          |
```

> | 🟢 SEV 5       | A low-level design inconsistency or minor issue that slightly affects usability or aesthetic, with minimal impact on users.                                                                                                                                                                                                                                     | Slight color deviations, minor typographical errors, or small spacing inconsistencies that do not impact functionality.                                                                                                                                                                                                                                                                                                                                                                                            |

Author: ```suggestion
| 🟢 SEV 5            | A low-level design inconsistency or minor issue that slightly affects usability or aesthetic, with minimal impact on users.                                                                                                                                                                                                                                     | Slight color deviations, minor typographical errors, or small spacing inconsistencies that do not impact functionality.                                                                                                                                                                                                                                                                                                                                                                                            |
```
